### PR TITLE
Issue a deprecation warning if a projects directory doesn't exist or can't be written to

### LINKED ISSUE #25531: Warn if project in settings.gradle has a non existent project directory
### Expected Behavior

If I have `settings.gradle` file that has
```
include ':doesntexist'
```

and there is no `doesntexist/` directory, Gradle should warn the user about it.

Gradle has implicit projectDir based on the project path, e.g. `:foo:bar` -> `foo/bar`. Creating a project has has non-existent directory is 99% likely to be a mistake.

### Current Behavior (optional)

_No response_

### Context

We found that our developers kept making mistakes (e.g. typo in the project name) and they didn't realize that the project is not actually included. We added our own check for this in settings.gradle, but with configuration cache enabled on Gradle 8.1, these projectDir.exists() checks lead to treating the whole projectDir as input to the configuration cache. We'd like to remove our own check and have gradle do this checking for us in CC safe way.

### PR DESCRIPTION
**Pull request by cobexer**
Fixes #25531

### CODE REVIEW

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/AbstractConfigurationAttributesResolveIntegrationTest.groovy [hunk 1]
```diff
@@ -866,6 +881,8 @@ All of them match the consumer attributes:
 
     def "selects configuration when it has more attributes than the resolved configuration"() {
         given:
+        createDirs("a", "b")
+        createDirs("a", "b")
```
Comment (ov7a) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/AbstractConfigurationAttributesResolveIntegrationTest.groovy:885]: Duplicate line

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformCachingIntegrationTest.groovy [hunk 2]
```diff
@@ -177,6 +178,9 @@ class ArtifactTransformCachingIntegrationTest extends AbstractHttpDependencyReso
                 }
             }
         """ << withFileLibDependency("lib1.jar", projectDir)
+        projectDir.createDir("lib")
```
Comment (ov7a) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformCachingIntegrationTest.groovy:181]: Can you use `createDirs` here?

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformCachingIntegrationTest.groovy [hunk 3]
```diff
@@ -925,6 +931,9 @@ class ArtifactTransformCachingIntegrationTest extends AbstractHttpDependencyReso
         [projectDir1, projectDir2].each { dir ->
             def project = new BuildTestFile(dir, "root")
             project.with {
+                dir.createDir("lib")
```
Comment (ov7a) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformCachingIntegrationTest.groovy:934]: Can you use `createDirs` here?

File: subprojects/core/src/integTest/groovy/org/gradle/api/SettingsIncludeManyIntegrationTest.groovy [hunk 4]
```diff
@@ -33,6 +33,9 @@ class SettingsIncludeManyIntegrationTest extends AbstractIntegrationSpec {
 
     @Requires(UnitTestPreconditions.IsGroovy3)
     def "including over 250 projects is not possible via varargs in Groovy 3"() {
+        createDirs(projectNames.take(254).collect({ name ->
```
Comment (ov7a) [inline subprojects/core/src/integTest/groovy/org/gradle/api/SettingsIncludeManyIntegrationTest.groovy:36]: 💅 I think it would be a bit clearer to extract some common logic for this and similar tests in a helper method.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy [hunk 5]
```diff
@@ -782,6 +782,7 @@ class CachedCustomTaskExecutionIntegrationTest extends AbstractIntegrationSpec i
         def projectNames = GroovyCollections.combinations(('a'..'p'), ('a'..'p'), ('a'..'d'))*.join("")
         println "Running with ${projectNames.size()} projects"
         projectNames.each { projectName ->
+            createDirs(projectName as String)
```
Comment (ov7a) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CachedCustomTaskExecutionIntegrationTest.groovy:785]: 💅 consider changing the type of `projectNames` instead of casting here

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java [hunk 6]
```diff
@@ -152,7 +152,14 @@ private void validate(SettingsInternal settings) {
                 String suffix = buildPath == Path.ROOT ? "" : " (in build " + buildPath + ")";
                 throw new GradleException("'" + SettingsInternal.BUILD_SRC + "' cannot be used as a project name as it is a reserved name" + suffix);
             }
+            if (!project.getProjectDir().exists() || !project.getProjectDir().isDirectory() || !project.getProjectDir().canWrite()) {
+                String msgTemplate = "Configuring project '%s' without an existing directory is deprecated. The configured projectDirectory '%s' does not exist, can't be written to or is not a directory.";
```
Comment (ov7a) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java:156]: 💅 Consider extracting the body block into a static method or `msgTemplate` in a static field.
💅 Consider renaming `msgTemplate` to `warningMessage` or something without shrtng of lttrs.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractAutoTestedSamplesTest.groovy [hunk 7]
```diff
@@ -23,6 +23,17 @@ class AbstractAutoTestedSamplesTest extends AbstractIntegrationTest {
     def util = new AutoTestedSamplesUtil()
 
     void runSamplesFrom(String dir) {
+        // make sure all project directories exist
+        def settingsEvaluatedHook = file("settingsEvaluatedHook.gradle") << """
```
Comment (ov7a) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AbstractAutoTestedSamplesTest.groovy:27]: 👍

File: subprojects/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 8]
```diff
@@ -47,6 +47,34 @@ The `groovy-base` plugin is now responsible for configuring source and target co
 If you are using this task *without applying `grooy-base`*, you will have to manually set compatibility versions on these tasks.
 In general, the `groovy-base` plugin should be applied whenever working with Groovy language tasks.
 
+[[deprecated_missing_project_directory]]
+==== Including projects without an existing directory
+
+Gradle will warn if a project is added to the build where the associated `projectDir` does not exist or is not writable.
+Starting with version 9.0 Gradle will refuse to run builds where any project directory is missing or read-only.
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:54]:
```suggestion
Starting with version 9.0, Gradle will not run builds if a project directory is missing or read-only.
```

**Review by lkasso — APPROVED** [review]
Docs approved.

**Review by ov7a — APPROVED** [review]

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java [hunk 9]
```diff
@@ -152,7 +152,18 @@ private void validate(SettingsInternal settings) {
...
             }
+            if (!project.getProjectDir().exists() || !project.getProjectDir().isDirectory() || !project.getProjectDir().canWrite()) {
+                emitProjectDirectoryMissingWarning(project.getPath(), project.getProjectDir().toString());
+            }
         });
     }
-}
 
+    private static void emitProjectDirectoryMissingWarning(String projectPath, String projectDir) {
+        String template = "Configuring project '%s' without an existing directory is deprecated. The configured projectDirectory '%s' does not exist, can't be written to or is not a directory.";
```
Comment (eskatos) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultSettingsLoader.java:162]: 👍

File: subprojects/docs/src/snippets/buildlifecycle/projectEvaluateEvents/groovy/project-a/.gitkeepdir
Comment (eskatos) [inline subprojects/docs/src/snippets/buildlifecycle/projectEvaluateEvents/groovy/project-a/.gitkeepdir:1]: ❔ Why use a `.gitkeepdir` file in this snippet instead of an empty build script?

File: subprojects/docs/src/snippets/multiproject/customLayout/groovy/settings.gradle [hunk 10]
```diff
@@ -8,6 +8,6 @@ println project(':project-a').name
 // end::lookup-project[]
 
 // tag::change-project[]
-project(':project-a').projectDir = file('../my-project-a')
+project(':project-a').projectDir = file('custom/my-project-a')
```
Comment (eskatos) [inline subprojects/docs/src/snippets/multiproject/customLayout/groovy/settings.gradle:11]: 👍

File: subprojects/docs/src/snippets/tasks/addDependencyUsingPath/groovy/build.gradle [hunk 11]
```diff
@@ -1,16 +0,0 @@
-project('project-a') {
```
Comment (eskatos) [inline subprojects/docs/src/snippets/tasks/addDependencyUsingPath/groovy/build.gradle:1]: 👍

File: subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle [hunk 12]
```diff
@@ -11,6 +11,8 @@ dependencies {
     testImplementation('org.spockframework:spock-core:2.2-groovy-3.0') {
         exclude group: 'org.codehaus.groovy'
     }
+    testImplementation 'org.junit.jupiter:junit-jupiter:5.10.0'
+    testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
```
Comment (eskatos) [inline subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle:15]: ❔ Why was this change necessary?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/TestProjectInitiation.groovy [hunk 13]
```diff
@@ -294,4 +294,11 @@ trait TestProjectInitiation {
         "build.gradle.kts"
     }
 
+    List<TestFile> createDirs(String... dirs) {
+        dirs.collect({ name ->
+            TestFile tf = file(name)
+            tf.mkdir()
```
Comment (eskatos) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/TestProjectInitiation.groovy:300]: ❓ Shouldn't this be `mkdirs()` to handle path with multiple segments in the argument instead?

**Review by eskatos — COMMENTED** [review]
A few questions

**Review by octylFractal — APPROVED** [review]

File: subprojects/docs/src/snippets/buildlifecycle/projectEvaluateEvents/groovy/project-a/.gitkeepdir
Comment (cobexer) [reply subprojects/docs/src/snippets/buildlifecycle/projectEvaluateEvents/groovy/project-a/.gitkeepdir:1]: Because `settings.gradle` of that project does:

```groovy
project(':project-a').buildFileName = '../project-a.gradle'
```

Which means that the project directory has to exist, but the build file does not reside inside of it.

File: subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle (hunk 12 above)
Comment (cobexer) [reply subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle:15]: because of this: https://docs.gradle.org/current/userguide/upgrading_version_8.html#manually_declaring_dependencies

Why I'm the one noticing that is a different interesting question tough.

Comment (eskatos) [reply subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle:15]: Was the snippet test failing without your changes?

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/TestProjectInitiation.groovy (hunk 13 above)
Comment (cobexer) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/build/TestProjectInitiation.groovy:300]: The idea was that it might catch situations where the directory is created relative to the wrong parent... but now that I made all the changes to integration tests it would probably succeed in those cases anyway and instead of this method failing the deprecation warning would still fail the test... so lets change it

File: subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle (hunk 12 above)
Comment (cobexer) [reply subprojects/docs/src/snippets/testKit/gradleVersion/groovy/build.gradle:15]: The test is also failing on master locally and is not executed in CI because of https://github.com/gradle/gradle/blob/a773afd64882f5b28ae24f3df3eeb4fa9765b2e3/subprojects/docs/build.gradle#L642C16-L642C16 and that `:docs:docsTest` in CI is only running on Java 20?

@eskatos Do you think I should back this change out into a separate PR?

**Review by eskatos — APPROVED** [review]

**Comment by cobexer** [comment]
@bot-gradle test and merge
